# Aim:
Initial exploration of the datasets. See what is available and what shape they are in

Dataset source : https://data.gov.ie/dataset/dublinbikes-api

In [2]:
import requests
import urllib.parse
import pandas as pd
import datetime as dt
from time import sleep

# Import a collection of useful functions
import useful

## Get latest snapshot of data (near real-time)

In [3]:
snapshot_api_url = "https://data.smartdublin.ie/dublinbikes-api/last_snapshot/"

In [4]:
# Send off request for data
snapshot_response = requests.get(snapshot_api_url)
snapshot_response.json()

[{'id': 48532447,
  'harvest_time': '2022-04-21T19:50:02',
  'station_id': 2,
  'available_bike_stands': 18,
  'bike_stands': 20,
  'available_bikes': 2,
  'banking': False,
  'bonus': False,
  'last_update': '2022-04-21T19:46:49',
  'status': 'OPEN',
  'address': 'Blessington Street',
  'name': 'BLESSINGTON STREET',
  'latitude': '53.3568',
  'longitude': '-6.26814'},
 {'id': 48532467,
  'harvest_time': '2022-04-21T19:50:02',
  'station_id': 3,
  'available_bike_stands': 9,
  'bike_stands': 20,
  'available_bikes': 11,
  'banking': False,
  'bonus': False,
  'last_update': '2022-04-21T19:48:52',
  'status': 'OPEN',
  'address': 'Bolton Street',
  'name': 'BOLTON STREET',
  'latitude': '53.3512',
  'longitude': '-6.26986'},
 {'id': 48532408,
  'harvest_time': '2022-04-21T19:50:02',
  'station_id': 4,
  'available_bike_stands': 16,
  'bike_stands': 20,
  'available_bikes': 4,
  'banking': False,
  'bonus': False,
  'last_update': '2022-04-21T19:49:00',
  'status': 'OPEN',
  'address': '

## Read in historical data
Downloaded from https://data.smartdublin.ie/dataset/dublinbikes-api

In [5]:
# Combine the dataframe together into one
data1 = pd.read_csv('data/dublinbikes_20190401_20190701.csv')
data2 = pd.read_csv('data/dublinbikes_20190701_20191001.csv')
data3 = pd.read_csv('data/dublinbikes_20191001_20200101.csv')
data4 = pd.read_csv('data/dublinbikes_20200101_20200401.csv')

#merge data into one table
df = pd.concat([data1, data2, data3, data4])

## What does the data consist of?

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10057161 entries, 0 to 2228277
Data columns (total 11 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   STATION ID             int64  
 1   TIME                   object 
 2   LAST UPDATED           object 
 3   NAME                   object 
 4   BIKE STANDS            int64  
 5   AVAILABLE BIKE STANDS  int64  
 6   AVAILABLE BIKES        int64  
 7   STATUS                 object 
 8   ADDRESS                object 
 9   LATITUDE               float64
 10  LONGITUDE              float64
dtypes: float64(2), int64(4), object(5)
memory usage: 920.8+ MB


In [7]:
df.head()

STATION ID                 TIME         LAST UPDATED                NAME  \
0           2  2019-04-01 00:00:06  2019-03-31 23:52:05  BLESSINGTON STREET   
1           2  2019-04-01 00:05:04  2019-04-01 00:02:13  BLESSINGTON STREET   
2           2  2019-04-01 00:10:04  2019-04-01 00:02:13  BLESSINGTON STREET   
3           2  2019-04-01 00:15:03  2019-04-01 00:12:21  BLESSINGTON STREET   
4           2  2019-04-01 00:20:03  2019-04-01 00:12:21  BLESSINGTON STREET   

   BIKE STANDS  AVAILABLE BIKE STANDS  AVAILABLE BIKES STATUS  \
0           20                     14                6   Open   
1           20                     14                6   Open   
2           20                     14                6   Open   
3           20                     14                6   Open   
4           20                     14                6   Open   

              ADDRESS  LATITUDE  LONGITUDE  
0  Blessington Street  53.35677   -6.26814  
1  Blessington Street  53.35677   -6.26814  
2  Blessington Street  53.35677   -6.26814  
3  Blessington Street  53.35677   -6.26814  
4  Blessington Street  53.35677   -6.26814

In [8]:
df.describe()

STATION ID   BIKE STANDS  AVAILABLE BIKE STANDS  AVAILABLE BIKES  \
count  1.005716e+07  1.005716e+07           1.005716e+07     1.005716e+07   
mean   5.920914e+01  3.206656e+01           2.016238e+01     1.154167e+01   
std    3.315855e+01  7.615782e+00           1.214401e+01     1.106513e+01   
min    2.000000e+00  1.600000e+01           0.000000e+00     0.000000e+00   
25%    3.100000e+01  2.900000e+01           1.000000e+01     2.000000e+00   
50%    5.900000e+01  3.000000e+01           2.000000e+01     8.000000e+00   
75%    8.800000e+01  4.000000e+01           3.000000e+01     1.900000e+01   
max    1.170000e+02  4.000000e+01           4.400000e+01     4.300000e+01   

           LATITUDE     LONGITUDE  
count  1.005716e+07  1.005716e+07  
mean   5.334552e+01 -6.264903e+00  
std    7.559545e-03  1.813792e-02  
min    5.333009e+01 -6.310015e+00  
25%    5.333976e+01 -6.277167e+00  
50%    5.334520e+01 -6.263371e+00  
75%    5.335093e+01 -6.251988e+00  
max    5.335997e+01 -6.230852e+00

__What about the frequency of values?__<br>
the pandas describe() function is useful for numerical columns, but doesn't tell us anything about categorical columns. Therefore, use frequency counts to explore the data columns more, particularly the categorical columns

In [9]:
useful.get_freq_counts(df)

STATION ID
115 unique values
Top 10
42    90771
18    90769
48    90767
81    90766
23    90765
31    90762
90    90761
45    90760
5     90760
91    90760
Name: STATION ID, dtype: int64
Bottom 10
116     9781
117    18109
46     20264
70     26119
14     59553
60     59553
35     68618
98     90212
11     90518
53     90523
Name: STATION ID, dtype: int64
TIME
90315 unique values
Top 10
2019-10-01 06:15:02    222
2019-10-01 11:50:03    222
2019-10-01 12:00:02    222
2019-10-01 12:05:02    222
2019-10-01 12:10:02    222
2019-10-01 12:15:02    222
2019-10-01 12:20:02    222
2019-10-01 12:25:02    222
2019-10-01 12:30:03    222
2019-10-01 12:35:03    222
Name: TIME, dtype: int64
Bottom 10
2020-01-25 21:45:02     4
2019-12-31 10:35:02     6
2020-01-24 04:10:02     7
2019-12-12 18:15:02     7
2019-12-10 13:55:02    12
2020-01-04 01:45:03    15
2020-01-25 21:30:03    25
2020-01-25 21:35:02    25
2020-01-25 21:40:02    25
2020-01-25 21:25:02    25
Name: TIME, dtype: int64
LAST UPDATED
5210060

### Take a closer look at the 'STATION ID' column

In [10]:
print(data1['STATION ID'].nunique())
print(data2['STATION ID'].nunique())
print(data3['STATION ID'].nunique())
print(data4['STATION ID'].nunique())

113
112
111
110


The number of stations in operation change over the course of the data set

### Take a closer look at the 'TIME' column

In [11]:
type(df.TIME[0])

pandas.core.series.Series

In [12]:
time_dates = list(set(map(lambda ind: ind.split(" ")[0], df.TIME)))
print(len(time_dates))
time_dates

325


['2020-03-18',
 '2020-02-12',
 '2019-11-25',
 '2019-05-01',
 '2019-04-16',
 '2020-02-11',
 '2019-11-26',
 '2020-03-10',
 '2019-10-15',
 '2020-03-05',
 '2019-11-12',
 '2019-09-05',
 '2019-08-14',
 '2019-05-28',
 '2019-05-19',
 '2019-07-21',
 '2020-02-25',
 '2019-10-07',
 '2019-09-02',
 '2019-12-01',
 '2019-10-24',
 '2019-12-14',
 '2020-01-31',
 '2019-12-22',
 '2019-05-17',
 '2019-05-10',
 '2019-05-14',
 '2019-11-02',
 '2019-11-18',
 '2019-12-02',
 '2019-12-17',
 '2020-03-03',
 '2019-10-21',
 '2019-09-12',
 '2019-08-27',
 '2019-09-20',
 '2019-11-17',
 '2019-04-09',
 '2019-08-28',
 '2019-07-01',
 '2020-02-10',
 '2020-03-15',
 '2019-11-22',
 '2019-10-25',
 '2019-09-15',
 '2020-03-21',
 '2019-07-25',
 '2019-07-09',
 '2020-02-28',
 '2019-12-07',
 '2020-03-31',
 '2019-05-09',
 '2019-10-03',
 '2020-01-29',
 '2019-10-18',
 '2019-10-01',
 '2019-07-15',
 '2020-03-30',
 '2019-12-11',
 '2019-12-23',
 '2019-09-08',
 '2019-05-25',
 '2019-11-30',
 '2020-02-20',
 '2020-03-22',
 '2019-05-11',
 '2019-08-

We have 325 unique days for a year's worth of data. We would expect 365 days

In [13]:
months = sorted(list(map(lambda ind: ind[:-3], time_dates)))
pd.Series(months).value_counts()

2019-05    31
2019-07    31
2019-08    31
2019-10    31
2020-03    31
2019-04    30
2019-09    30
2019-11    30
2020-02    29
2019-12    23
2020-01    15
2019-06    12
2020-04     1
dtype: int64

Possibly missing data for 18 days in 2019-06, 8 days in 2019-12 and 16 days in 2020-01

In [14]:
time_times = list(set(map(lambda ind: ind.split(" ")[1], df['TIME'])))
print(len(time_times))
time_times

1830


['00:35:02',
 '15:45:03',
 '20:05:27',
 '10:55:02',
 '17:50:27',
 '09:55:03',
 '23:40:03',
 '22:00:05',
 '09:00:03',
 '06:20:03',
 '08:50:04',
 '10:35:01',
 '18:25:01',
 '07:20:01',
 '13:50:07',
 '15:55:27',
 '02:50:03',
 '19:50:04',
 '01:25:02',
 '08:20:02',
 '22:00:04',
 '22:00:02',
 '22:15:04',
 '23:00:06',
 '08:40:06',
 '15:00:05',
 '01:55:27',
 '13:20:03',
 '21:35:27',
 '12:55:04',
 '20:30:05',
 '13:15:03',
 '20:20:02',
 '12:50:04',
 '19:00:04',
 '07:40:03',
 '01:40:27',
 '19:10:28',
 '05:55:03',
 '14:25:02',
 '13:35:04',
 '08:30:04',
 '17:40:07',
 '11:00:02',
 '06:50:03',
 '21:10:27',
 '18:55:03',
 '07:00:02',
 '19:30:27',
 '11:35:03',
 '12:00:02',
 '20:50:04',
 '02:30:27',
 '08:30:27',
 '15:50:05',
 '15:10:04',
 '15:10:05',
 '03:55:01',
 '08:50:02',
 '06:50:02',
 '11:55:02',
 '05:35:27',
 '01:50:04',
 '09:00:04',
 '19:50:03',
 '19:55:27',
 '08:05:04',
 '14:20:02',
 '01:55:02',
 '00:25:04',
 '13:30:01',
 '06:45:27',
 '22:20:05',
 '00:05:28',
 '03:20:27',
 '09:10:04',
 '22:25:28',

In [15]:
sorted(time_times)

['00:00:01',
 '00:00:02',
 '00:00:03',
 '00:00:04',
 '00:00:05',
 '00:00:06',
 '00:00:07',
 '00:00:08',
 '00:02:09',
 '00:05:01',
 '00:05:02',
 '00:05:03',
 '00:05:04',
 '00:05:07',
 '00:05:08',
 '00:05:27',
 '00:05:28',
 '00:05:30',
 '00:07:10',
 '00:10:02',
 '00:10:03',
 '00:10:04',
 '00:10:27',
 '00:12:09',
 '00:15:01',
 '00:15:02',
 '00:15:03',
 '00:15:04',
 '00:15:27',
 '00:20:02',
 '00:20:03',
 '00:20:04',
 '00:20:05',
 '00:20:27',
 '00:25:01',
 '00:25:02',
 '00:25:03',
 '00:25:04',
 '00:25:27',
 '00:30:01',
 '00:30:02',
 '00:30:03',
 '00:30:04',
 '00:30:05',
 '00:30:27',
 '00:35:01',
 '00:35:02',
 '00:35:03',
 '00:35:04',
 '00:35:27',
 '00:40:01',
 '00:40:02',
 '00:40:03',
 '00:40:04',
 '00:45:01',
 '00:45:02',
 '00:45:03',
 '00:45:04',
 '00:45:27',
 '00:50:01',
 '00:50:02',
 '00:50:03',
 '00:50:04',
 '00:50:28',
 '00:50:29',
 '00:55:01',
 '00:55:02',
 '00:55:03',
 '00:55:04',
 '00:55:27',
 '00:55:28',
 '01:00:01',
 '01:00:02',
 '01:00:03',
 '01:00:04',
 '01:00:05',
 '01:00:06',

Strip it down to HH:MM

In [16]:
time_times = list(set(map(lambda ind: ind.split(" ")[1][:5], df['TIME'])))
print(len(time_times))
time_times

306


['11:25',
 '04:20',
 '11:45',
 '21:40',
 '16:45',
 '14:15',
 '02:50',
 '16:35',
 '12:20',
 '02:30',
 '05:40',
 '20:00',
 '01:45',
 '18:25',
 '21:10',
 '08:55',
 '22:50',
 '15:30',
 '18:55',
 '00:00',
 '23:25',
 '10:30',
 '19:05',
 '21:05',
 '07:45',
 '11:05',
 '21:15',
 '22:25',
 '23:32',
 '02:40',
 '05:30',
 '08:40',
 '07:55',
 '03:40',
 '01:50',
 '05:00',
 '10:00',
 '15:15',
 '17:05',
 '17:30',
 '16:50',
 '02:10',
 '09:15',
 '14:25',
 '07:35',
 '16:00',
 '12:15',
 '12:35',
 '09:55',
 '13:15',
 '08:25',
 '21:45',
 '01:15',
 '17:15',
 '21:00',
 '10:10',
 '04:40',
 '22:55',
 '22:35',
 '14:10',
 '18:15',
 '06:30',
 '19:45',
 '02:45',
 '04:35',
 '00:45',
 '18:45',
 '14:50',
 '17:35',
 '09:25',
 '20:20',
 '20:10',
 '09:00',
 '20:45',
 '10:25',
 '06:55',
 '23:05',
 '09:05',
 '02:20',
 '19:20',
 '06:35',
 '11:30',
 '12:40',
 '08:00',
 '06:50',
 '04:25',
 '02:00',
 '10:35',
 '22:45',
 '00:20',
 '05:25',
 '16:25',
 '13:40',
 '17:20',
 '13:20',
 '23:50',
 '03:10',
 '14:30',
 '22:30',
 '08:52',


Expected 288 unique timepoints as they are taken every 5 mins so 12 * 24 = 288. However, 306 unique values are returned

In [17]:
mins = list(set(map(lambda ind: ind.split(" ")[1][3:5], df['TIME'])))
print(len(mins))
mins

25


['30',
 '42',
 '20',
 '50',
 '41',
 '55',
 '35',
 '07',
 '15',
 '40',
 '12',
 '00',
 '37',
 '47',
 '02',
 '57',
 '10',
 '45',
 '52',
 '25',
 '06',
 '05',
 '22',
 '31',
 '32']

Unlike the data of later years, the earlier data is not always recorded at 5 min intervals

### Take a closer look at the 'LAST UPDATED' column

In [18]:
type(df['LAST UPDATED'][0])

pandas.core.series.Series

In [19]:
last_updated_dates = list(set(map(lambda ind: ind.split(" ")[0], df['LAST UPDATED'])))
print(len(last_updated_dates))
last_updated_dates

330


['2020-03-18',
 '2020-02-12',
 '2019-11-25',
 '2019-05-01',
 '2019-04-16',
 '2020-02-11',
 '2019-11-26',
 '2020-03-10',
 '2019-10-15',
 '2020-03-05',
 '2019-11-12',
 '2019-09-05',
 '2019-08-14',
 '2019-05-28',
 '2019-05-19',
 '2019-07-21',
 '2020-02-25',
 '2019-10-07',
 '2019-09-02',
 '2019-12-01',
 '2019-10-24',
 '2019-12-14',
 '2020-01-31',
 '2019-12-22',
 '2019-05-17',
 '2019-05-10',
 '2019-05-14',
 '2019-11-02',
 '2019-11-18',
 '2020-03-03',
 '2019-12-02',
 '2019-12-17',
 '2019-10-21',
 '2019-09-12',
 '2019-08-27',
 '2020-01-22',
 '2019-09-20',
 '2020-01-12',
 '2019-11-17',
 '2019-04-09',
 '2019-08-28',
 '2019-07-01',
 '2020-02-10',
 '2020-03-15',
 '2019-11-22',
 '2019-10-25',
 '2019-09-15',
 '2020-03-21',
 '2019-07-25',
 '2019-07-09',
 '2020-02-28',
 '2019-12-07',
 '2020-03-31',
 '2019-05-09',
 '2019-10-03',
 '2020-01-29',
 '2019-10-18',
 '2019-10-01',
 '2019-07-15',
 '2020-03-30',
 '2019-12-11',
 '2019-12-23',
 '2019-09-08',
 '2019-05-25',
 '2019-11-30',
 '2020-02-20',
 '2020-03-

In [20]:
last_updated_times = list(set(map(lambda ind: ind.split(" ")[1], df['LAST UPDATED'])))
print(len(last_updated_times))
last_updated_times

86400


['16:52:37',
 '00:13:10',
 '01:15:20',
 '05:19:44',
 '09:35:11',
 '01:13:37',
 '14:15:39',
 '10:59:15',
 '22:11:59',
 '19:25:34',
 '12:29:25',
 '19:16:35',
 '11:09:57',
 '19:22:44',
 '19:02:25',
 '06:12:38',
 '13:47:48',
 '11:34:57',
 '20:59:41',
 '17:00:19',
 '22:11:21',
 '06:43:39',
 '19:42:41',
 '18:37:28',
 '21:37:04',
 '08:07:14',
 '11:22:34',
 '10:46:07',
 '03:57:58',
 '02:33:28',
 '07:21:43',
 '15:25:09',
 '01:43:14',
 '04:40:38',
 '11:37:28',
 '13:45:56',
 '02:18:41',
 '14:18:07',
 '15:19:27',
 '22:16:48',
 '00:02:15',
 '03:18:52',
 '15:26:08',
 '02:54:54',
 '18:56:12',
 '00:39:16',
 '10:30:42',
 '13:11:32',
 '02:06:36',
 '00:09:00',
 '21:36:57',
 '20:51:09',
 '21:00:44',
 '20:11:18',
 '17:51:53',
 '11:20:38',
 '05:24:03',
 '13:26:52',
 '20:05:41',
 '13:52:58',
 '16:23:35',
 '11:43:40',
 '17:35:16',
 '13:51:22',
 '03:57:52',
 '19:33:02',
 '06:49:04',
 '08:05:33',
 '13:44:24',
 '08:34:55',
 '20:43:03',
 '11:09:29',
 '04:28:53',
 '05:36:34',
 '06:44:51',
 '14:00:18',
 '18:37:12',

### Take a closer look at 'BIKE STANDS', 'AVAILABLE BIKE STANDS' and 'AVAILABLE BIKES'

In [21]:
tmp = df.copy()
tmp['bike_stand_diff'] = tmp['BIKE STANDS'] - tmp['AVAILABLE BIKE STANDS']
tmp[tmp.bike_stand_diff != 0].head()

STATION ID                 TIME         LAST UPDATED                NAME  \
0           2  2019-04-01 00:00:06  2019-03-31 23:52:05  BLESSINGTON STREET   
1           2  2019-04-01 00:05:04  2019-04-01 00:02:13  BLESSINGTON STREET   
2           2  2019-04-01 00:10:04  2019-04-01 00:02:13  BLESSINGTON STREET   
3           2  2019-04-01 00:15:03  2019-04-01 00:12:21  BLESSINGTON STREET   
4           2  2019-04-01 00:20:03  2019-04-01 00:12:21  BLESSINGTON STREET   

   BIKE STANDS  AVAILABLE BIKE STANDS  AVAILABLE BIKES STATUS  \
0           20                     14                6   Open   
1           20                     14                6   Open   
2           20                     14                6   Open   
3           20                     14                6   Open   
4           20                     14                6   Open   

              ADDRESS  LATITUDE  LONGITUDE  bike_stand_diff  
0  Blessington Street  53.35677   -6.26814                6  
1  Blessington Street  53.35677   -6.26814                6  
2  Blessington Street  53.35677   -6.26814                6  
3  Blessington Street  53.35677   -6.26814                6  
4  Blessington Street  53.35677   -6.26814                6

In [22]:
tmp[tmp['AVAILABLE BIKE STANDS']> tmp['AVAILABLE BIKES']].head()

STATION ID                 TIME         LAST UPDATED                NAME  \
0           2  2019-04-01 00:00:06  2019-03-31 23:52:05  BLESSINGTON STREET   
1           2  2019-04-01 00:05:04  2019-04-01 00:02:13  BLESSINGTON STREET   
2           2  2019-04-01 00:10:04  2019-04-01 00:02:13  BLESSINGTON STREET   
3           2  2019-04-01 00:15:03  2019-04-01 00:12:21  BLESSINGTON STREET   
4           2  2019-04-01 00:20:03  2019-04-01 00:12:21  BLESSINGTON STREET   

   BIKE STANDS  AVAILABLE BIKE STANDS  AVAILABLE BIKES STATUS  \
0           20                     14                6   Open   
1           20                     14                6   Open   
2           20                     14                6   Open   
3           20                     14                6   Open   
4           20                     14                6   Open   

              ADDRESS  LATITUDE  LONGITUDE  bike_stand_diff  
0  Blessington Street  53.35677   -6.26814                6  
1  Blessington Street  53.35677   -6.26814                6  
2  Blessington Street  53.35677   -6.26814                6  
3  Blessington Street  53.35677   -6.26814                6  
4  Blessington Street  53.35677   -6.26814                6

In [23]:
tmp['avail_stands_and_bikes'] = tmp['AVAILABLE BIKE STANDS'] + tmp['AVAILABLE BIKES']
tmp[tmp['BIKE STANDS'] != tmp.avail_stands_and_bikes]

STATION ID                 TIME         LAST UPDATED  \
382               3  2019-04-01 07:55:03  2019-04-01 07:53:58   
658               4  2019-04-01 07:00:07  2019-04-01 06:59:09   
673               4  2019-04-01 08:15:03  2019-04-01 08:10:29   
674               4  2019-04-01 08:20:02  2019-04-01 08:10:29   
1050              5  2019-04-01 15:45:03  2019-04-01 15:43:49   
...             ...                  ...                  ...   
2227533         115  2020-04-01 09:55:02  2020-04-01 09:52:47   
2227534         115  2020-04-01 10:00:03  2020-04-01 09:52:47   
2227535         115  2020-04-01 10:05:02  2020-04-01 10:03:46   
2227536         115  2020-04-01 10:10:03  2020-04-01 10:03:46   
2227537         115  2020-04-01 10:15:03  2020-04-01 10:13:55   

                     NAME  BIKE STANDS  AVAILABLE BIKE STANDS  \
382         BOLTON STREET           20                     19   
658          GREEK STREET           20                     19   
673          GREEK STREET           20                     19   
674          GREEK STREET           20                     19   
1050     CHARLEMONT PLACE           40                     27   
...                   ...          ...                    ...   
2227533  KILLARNEY STREET           30                      9   
2227534  KILLARNEY STREET           30                      9   
2227535  KILLARNEY STREET           30                     10   
2227536  KILLARNEY STREET           30                     10   
2227537  KILLARNEY STREET           30                     10   

         AVAILABLE BIKES STATUS            ADDRESS   LATITUDE  LONGITUDE  \
382                    0   Open      Bolton Street  53.351181  -6.269859   
658                    0   Open       Greek Street  53.346874  -6.272976   
673                    0   Open       Greek Street  53.346874  -6.272976   
674                    0   Open       Greek Street  53.346874  -6.272976   
1050                  12   Open  Charlemont Street  53.330662  -6.260177   
...                  ...    ...                ...        ...        ...   
2227533               20   Open   Killarney Street  53.354843  -6.247579   
2227534               20   Open   Killarney Street  53.354843  -6.247579   
2227535               19   Open   Killarney Street  53.354843  -6.247579   
2227536               19   Open   Killarney Street  53.354843  -6.247579   
2227537               19   Open   Killarney Street  53.354843  -6.247579   

         bike_stand_diff  avail_stands_and_bikes  
382                    1                      19  
658                    1                      19  
673                    1                      19  
674                    1                      19  
1050                  13                      39  
...                  ...                     ...  
2227533               21                      29  
2227534               21                      29  
2227535               20                      29  
2227536               20                      29  
2227537               20                      29  

[515611 rows x 13 columns]

Do we ever run out of available bike stands?

In [24]:
print(len(tmp[tmp['AVAILABLE BIKE STANDS'] == 0]))
tmp[tmp['AVAILABLE BIKE STANDS'] == 0].head()

594436


STATION ID                 TIME         LAST UPDATED           NAME  \
287           3  2019-04-01 00:00:06  2019-03-31 23:52:39  BOLTON STREET   
288           3  2019-04-01 00:05:04  2019-04-01 00:02:48  BOLTON STREET   
289           3  2019-04-01 00:10:04  2019-04-01 00:02:48  BOLTON STREET   
290           3  2019-04-01 00:15:03  2019-04-01 00:12:56  BOLTON STREET   
291           3  2019-04-01 00:20:03  2019-04-01 00:12:56  BOLTON STREET   

     BIKE STANDS  AVAILABLE BIKE STANDS  AVAILABLE BIKES STATUS  \
287           20                      0               20   Open   
288           20                      0               20   Open   
289           20                      0               20   Open   
290           20                      0               20   Open   
291           20                      0               20   Open   

           ADDRESS   LATITUDE  LONGITUDE  bike_stand_diff  \
287  Bolton Street  53.351181  -6.269859               20   
288  Bolton Street  53.351181  -6.269859               20   
289  Bolton Street  53.351181  -6.269859               20   
290  Bolton Street  53.351181  -6.269859               20   
291  Bolton Street  53.351181  -6.269859               20   

     avail_stands_and_bikes  
287                      20  
288                      20  
289                      20  
290                      20  
291                      20

Do we ever run out of available bikes?

In [25]:
print(len(tmp[tmp['AVAILABLE BIKES'] == 0]))
tmp[tmp['AVAILABLE BIKES'] == 0].head()

1688770


STATION ID                 TIME         LAST UPDATED                NAME  \
70           2  2019-04-01 05:50:03  2019-04-01 05:48:58  BLESSINGTON STREET   
71           2  2019-04-01 05:55:03  2019-04-01 05:48:58  BLESSINGTON STREET   
72           2  2019-04-01 06:00:04  2019-04-01 05:48:58  BLESSINGTON STREET   
73           2  2019-04-01 06:05:03  2019-04-01 05:59:06  BLESSINGTON STREET   
76           2  2019-04-01 06:20:28  2019-04-01 06:19:29  BLESSINGTON STREET   

    BIKE STANDS  AVAILABLE BIKE STANDS  AVAILABLE BIKES STATUS  \
70           20                     20                0   Open   
71           20                     20                0   Open   
72           20                     20                0   Open   
73           20                     20                0   Open   
76           20                     20                0   Open   

               ADDRESS  LATITUDE  LONGITUDE  bike_stand_diff  \
70  Blessington Street  53.35677   -6.26814                0   
71  Blessington Street  53.35677   -6.26814                0   
72  Blessington Street  53.35677   -6.26814                0   
73  Blessington Street  53.35677   -6.26814                0   
76  Blessington Street  53.35677   -6.26814                0   

    avail_stands_and_bikes  
70                      20  
71                      20  
72                      20  
73                      20  
76                      20

### Take a closer look at 'STATUS'

In [26]:
closed_df = df[df.STATUS != 'Open'].copy()
closed_df[['STATION ID', 'STATUS']].groupby('STATION ID').count()


STATUS
STATION ID        
7                1
14            2655
16             605
23             674
25            2470
26            2029
35           60376
46            3723
56             747
60            1199
70             470
73            1951
89            1267
98           15463
113           2446
114             44

## Initial observations:

 - 115 unique stations - For these we stations, we have station ids, names, addresses and geographical locations. We should consider these as static pieces of information i.e. that they are unlikely to change -> take these out and create a separate lookup table with these. However the number of stations that are active during the time frame under study does change.
 
 
 - 'BIKE STANDS' is the number of stands associated with a station.'AVAILABLE BIKE STANDS' are the number of bike stands that are empty or have no bike in them. 'AVAILABLE BIKES' is the number of bikes currently available for use at the station. In general, 'BIKE STANDS' = 'AVAILABLE BIKES' + 'AVAILABLE BIKE STANDS', however that is not always the case. Sometimes, 'BIKE STANDS' > 'AVAILABLE BIKES' + 'AVAILABLE BIKE STANDS' i.e. assume that some bike stands are out of order. How often does it happen? Could we consider the number of bike stands associated with a station relatively static? 
 
 
 - 'LAST UPDATED' and 'TIME' are both columns that consist of values that are a timestamp in the form of a string consisting of a date in YYYY-MM-DD format separated by a " " from a time in the format of HH:MM:SS. 'TIME' contains data from 5 min intervals for a given hour
 - Would be useful to break 'TIME' into two columns - 'date' and 'time'. Also 'day_of_week' would be useful
 
 
- Possibly missing data for 18 days in 2019-06, 8 days in 2019-12 and 16 days in 2020-01. Also, unlike the data of later years, the earlier data is not always recorded at 5 min intervals => possible discontinuous data that we would need to consider if we want to forecast demand

 
- 'STATUS': 15 different stations had a status of closed during the given time frame
 